lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,959 @@
-    }
-    if (tpe.getValue() == "companion")
-      op->setAttr("lowerToBind", BoolAttr::get(op.getContext(), true));
-  }
+  auto *op = value.getDefiningOp();
+  if (InstanceOp instance = dyn_cast<InstanceOp>(op))
+    return instance.getReferencedModule();
 
-  visitUnhandledDecl(op);
+  return op->getParentOfType<FModuleOp>();
 }
 
+/// This method contains the business logic of this pass.
 void GrandCentralPass::runOnOperation() {
+  LLVM_DEBUG(llvm::dbgs() << "===- Running Grand Central Views/Interface Pass "
+                             "-----------------------------===\n");
+
   CircuitOp circuitOp = getOperation();
 
   AnnotationSet annotations(circuitOp);
   if (annotations.empty())
-    return;
+    return markAllAnalysesPreserved();
+
+  // Pull out the extraction info if it exists.
+  bool removalError = false, workToDo = false;
+  annotations.removeAnnotations([&](Annotation anno) {
+    if (anno.isClass(
+            "sifive.enterprise.grandcentral.ExtractGrandCentralAnnotation")) {
+      if (maybeExtractInfo.hasValue()) {
+        emitCircuitError("more than one 'ExtractGrandCentralAnnotation' was "
+                         "found, but exactly one must be provided");
+        removalError = true;
+        return false;
+      }
+
+      maybeExtractInfo = {anno.getMember<StringAttr>("directory").getValue(),
+                          anno.getMember<StringAttr>("filename").getValue()};
+      return false;
+    }
+    workToDo |=
+        anno.isClass("sifive.enterprise.grandcentral.AugmentedBundleType");
+    return false;
+  });
+
+  if (removalError)
+    return signalPassFailure();
+
+  // Exit immediately if no annotations indicative of interfaces that need to be
+  // built exist.
+  if (!workToDo)
+    return markAllAnalysesPreserved();
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "Extraction Info:\n";
+    if (maybeExtractInfo)
+      llvm::dbgs() << "  directory: " << maybeExtractInfo.getValue().directory
+                   << "\n"
+                   << "  filename: " << maybeExtractInfo.getValue().bindFilename
+                   << "\n";
+    else
+      llvm::dbgs() << "  <none>\n";
+  });
 
   // Setup the builder to create ops _inside the FIRRTL circuit_.  This is
   // necessary because interfaces and interface instances are created.
-  // Instances link to their definitions via symbols and we don't want to break
-  // this.
+  // Instances link to their definitions via symbols and we don't want to
+  // break this.
   auto builder = OpBuilder::atBlockEnd(circuitOp.getBody());
 
-  // Utility that acts like emitOpError, but does _not_ include a note.  The
-  // note in emitOpError includes the entire op which means the **ENTIRE**
-  // FIRRTL circuit.  This doesn't communicate anything useful to the user other
-  // than flooding their terminal.
-  auto emitCircuitError = [&circuitOp](StringRef message = {}) {
-    return emitError(circuitOp.getLoc(), message);
-  };
+  SymbolTable symbolTable(circuitOp);
```

Comment:
Neat, this might go well with that `getReferencedModule` TODO further up :grimacing:

Reply:
Fixed in the (forthcoming) rebase.